src/c++/perf_analyzer/client_backend/openai/openai_client.h

```diff
@@ -0,0 +1,186 @@
+      bool is_null, const std::string& request_id)
+      : http_code_(http_code),
+        serialized_response_(std::move(serialized_response)),
+        is_final_(is_final), is_null_(is_null), request_id_(request_id)
+  {
+  }
+  virtual ~ChatCompletionResult() = default;
+
+  /// Get the id of the request which generated this response.
+  /// \param id Returns the request id that generated the result.
+  /// \return Error object indicating success or failure.
+  Error Id(std::string* id) const override
+  {
+    *id = request_id_;
+    return Error::Success;
+  }
+
+
+  /// Returns the status of the request.
+  /// \return Error object indicating the success or failure of the
+  /// request.
+  Error RequestStatus() const override
+  {
+    if ((http_code_ >= 400) && (http_code_ <= 599)) {
+      return Error(
+          "OpenAI response returns HTTP code" + std::to_string(http_code_));
+    }
+    return Error::Success;
+  }
+
+  /// Returns the raw data of the output.
+  /// \return Error object indicating the success or failure of the
+  /// request.
+  Error RawData(
+      const std::string& output_name, const uint8_t** buf,
+      size_t* byte_size) const override
+  {
+    // [FIXME] disregard "output_name" which is not compatible to
+    // OpenAI protocol
+    *buf = reinterpret_cast<const uint8_t*>(serialized_response_.c_str());
+    *byte_size = serialized_response_.size();
+    return Error::Success;
+  }
+
+  /// Get final response bool for this response.
+  /// \return Error object indicating the success or failure.
+  Error IsFinalResponse(bool* is_final_response) const override
+  {
+    *is_final_response = is_final_;
+    return Error::Success;
+  };
+
+  /// Get null response bool for this response.
+  /// \return Error object indicating the success or failure.
+  Error IsNullResponse(bool* is_null_response) const override
+  {
+    *is_null_response = is_null_;
+    return Error::Success;
+  };
+
+ private:
+  const uint32_t http_code_{200};
+  const std::string serialized_response_;
+  const bool is_final_{false};
+  const bool is_null_{false};
+  const std::string request_id_;
+};
+
+
+class ChatCompletionRequest : public HttpRequest {
+ public:
+  virtual ~ChatCompletionRequest() {}
+  ChatCompletionRequest(
+      std::function<void(HttpRequest*)>&& completion_callback,
+      std::function<void(InferResult*)>&& response_callback,
+      const std::string& request_id,
+      const bool verbose = false)
+      : HttpRequest(std::move(completion_callback), verbose),
+        response_callback_(std::move(response_callback)),
+        request_id_(request_id)
+  {
+  }
+  void SendResponse(bool is_final, bool is_null);
+  bool is_stream_{false};
+  std::function<void(InferResult*)> response_callback_{nullptr};
+  // The timers for infer request.
+  triton::client::RequestTimers timer_;
+  const std::string request_id_;
+};
+
+class ChatCompletionClient : public HttpClient {
```

Comment:
I don't think there is specialized logic with respect to chat completion endpoint. It does some checking for server-sent events which is a generic HTTP concept. I am open to other name for the class, but I am hesitated to just call it OpenAIClient since I haven't checked on how other OpenAI endpoints defines their transaction logic.